test/e2e/md_upgrades.go

```diff
@@ -0,0 +1,120 @@
+		input         MDUpgradesSpecInput
+		namespace     *corev1.Namespace
+		cancelWatches context.CancelFunc
+		cluster       *clusterv1.Cluster
+	)
+
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		input = inputGetter()
+		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(input.ArtifactFolder, 0755)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersionCurrent))
+		Expect(input.E2EConfig.Variables).To(HaveValidVersion(input.E2EConfig.GetVariable(KubernetesVersionCurrent)))
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersionPrevious))
+		Expect(input.E2EConfig.Variables).To(HaveValidVersion(input.E2EConfig.GetVariable(KubernetesVersionPrevious)))
+		Expect(input.E2EConfig.Variables).To(HaveKey(CNIPath))
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+	})
+
+	It("Should successfully upgrade Machines upon changes in relevant MachineDeployment fields", func() {
+
+		By("Creating a workload cluster")
+
+		var mds []*clusterv1.MachineDeployment
+		cluster, _, mds = clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
+			ClusterProxy: input.BootstrapClusterProxy,
+			ConfigCluster: clusterctl.ConfigClusterInput{
+				LogFolder:                filepath.Join(input.ArtifactFolder, "clusters", input.BootstrapClusterProxy.GetName()),
+				ClusterctlConfigPath:     input.ClusterctlConfigPath,
+				KubeconfigPath:           input.BootstrapClusterProxy.GetKubeconfigPath(),
+				InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+				Flavor:                   clusterctl.DefaultFlavor,
+				Namespace:                namespace.Name,
+				ClusterName:              fmt.Sprintf("cluster-%s", util.RandomString(6)),
+				KubernetesVersion:        input.E2EConfig.GetVariable(KubernetesVersionPrevious),
+				ControlPlaneMachineCount: pointer.Int64Ptr(1),
+				WorkerMachineCount:       pointer.Int64Ptr(1),
+			},
+			CNIManifestPath:              input.E2EConfig.GetVariable(CNIPath),
+			WaitForClusterIntervals:      input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(specName, "wait-control-plane"),
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
```

Comment:
Apologies if this has been asked already (haven't had the time to go through each comment), but could we reuse of of the other clusters or is it preferable to create a new one?

Reply:
To logically separate what each spec does, @fabriziopandini suggested to  make these different specs and not combine them into 1 spec, and I agree as this will make useability/debugging easier for providers when calling these specs.

Reply:
In terms of time, how long does it take to get a new workload cluster up and running?

Reply:
@vincepri this is a ginkgo trade-off.
At this stage, I'm favoring a clean separation between tests and relying on parallelism to keep test time under control.

Alternatives come with a downside
- option 1 conflating everything in a single `It` CONS: readability/reusability
- option 2 create the cluster once and have many `It` CONS: make sense only if each `It` is not dependent from the other CONS: doesn't really fit the use case we have met so far where everything is a unique sequence of operation

Reply:
Each spec should be able to reuse an existing cluster, or create one as a fallback mechanism.

Conflating all tests in one `It` statement is far from ideal, but if we allow specs to share a cluster, we could have some locking mechanics based on some simple conditions.

As an example, we should be able to run a MachineDeployment upgrade after the control plane has been created.

Regardless, this can be a follow-up improvement

Reply:
@vincepri I agree there is room for future improvements
but the last test run for this PR took 26 minutes, and we are still running on parallelism 2, so the baseline seems not so bad